File: 230. Kth Smallest Element in a BST.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @param {number} k
 * @return {number}
 */
var kthSmallest = function(root, k) {   // DFS of a BST will get a sorted array, so the nth time we visit the node, the node will be the nth smallest
    let result;
    (function dfs(node) {
        if (node.left) {
            dfs(node.left);
        }
        k--;
        if (k === 0) {
            result = node.val;
        }
        if (node.right) {
            dfs(node.right);
        }
    })(root);
    return result;
};


var kthSmallest = function(root, k) {  // Morris inOrder traversal
    let current = root;
    let previous;
    while(current) {
        if (current.left) {
            previous = current.left;
            while(previous.right && previous.right !== current) {
                previous = previous.right;   
            }
            if (previous.right) {
                previous.right = null;
                k--;
                if (k === 0) {
                    return current.val;   
                }
                current = current.right;
            } else {
                previous.right = current;
                current = current.left;
            }
        } else {
            k--;
            if (k === 0) {
                return current.val;   
            }
            current = current.right;
        }
    }
}

File: Golang/Tree/230.Kth-smallest-element-in-a-bst.go
package main

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func kthSmallest(root *TreeNode, k int) int {
	current := root
	stack := []*TreeNode{}
	var ret int
	for k > 0 {
		if current != nil {
			stack = append(stack, current)
			current = current.Left
		} else if len(stack) > 0 {
			node := stack[len(stack)-1]
			stack = stack[:len(stack)-1]
			ret = node.Val
			k--
			current = node.Right
		} else {
			break
		}
	}
	return ret
}
